v2/yarpctchannel/outbound.go

```diff
@@ -118,120 +114,146 @@ func callWithPeer(ctx context.Context, req *transport.Request, peer *tchannel.Pe
 		// TODO(abg): This will wrap IO errors while writing headers as encode
 		// errors. We should fix that.
-		return nil, errors.RequestHeadersEncodeError(req, err)
+		return nil, nil, yarpcencoding.RequestHeadersEncodeError(req, err)
 	}
 
-	if err := writeBody(req.Body, call); err != nil {
-		return nil, err
+	if err := writeBody(reqBody, call); err != nil {
+		return nil, nil, err
 	}
 
 	res := call.Response()
 	headers, err := readHeaders(format, res.Arg2Reader)
 	if err != nil {
 		if err, ok := err.(tchannel.SystemError); ok {
-			return nil, fromSystemError(err)
+			return nil, nil, fromSystemError(err)
 		}
 		// TODO(abg): This will wrap IO errors while reading headers as decode
 		// errors. We should fix that.
-		return nil, errors.ResponseHeadersDecodeError(req, err)
+		return nil, nil, yarpcencoding.ResponseHeadersDecodeError(req, err)
 	}
 
-	resBody, err := res.Arg3Reader()
+	arg3Reader, err := res.Arg3Reader()
 	if err != nil {
 		if err, ok := err.(tchannel.SystemError); ok {
-			return nil, fromSystemError(err)
+			return nil, nil, fromSystemError(err)
 		}
-		return nil, err
+		return nil, nil, err
+	}
+	resBody := &yarpc.Buffer{}
+	_, err = internaliopool.Copy(resBody, arg3Reader)
+	if err != nil {
+		return nil, nil, err
 	}
 
-	// service name match validation, return yarpcerrors.CodeInternal error if not match
+	// service name match validation, return yarpcerror.CodeInternal error if not match
 	if match, resSvcName := checkServiceMatchAndDeleteHeaderKey(req.Service, headers); !match {
-		return nil, yarpcerrors.InternalErrorf("service name sent from the request "+
+		return nil, nil, yarpcerror.InternalErrorf("service name sent from the request "+
 			"does not match the service name received in the response: sent %q, got: %q", req.Service, resSvcName)
 	}
 
-	return &transport.Response{
+	return &yarpc.Response{
 		Headers:          headers,
-		Body:             resBody,
 		ApplicationError: res.ApplicationError(),
-	}, getResponseErrorAndDeleteHeaderKeys(headers)
+	}, resBody, getResponseErrorAndDeleteHeaderKeys(headers)
 }
 
-func (o *Outbound) getPeerForRequest(ctx context.Context, treq *transport.Request) (*tchannelPeer, func(error), error) {
-	p, onFinish, err := o.chooser.Choose(ctx, treq)
-	if err != nil {
-		return nil, nil, err
+func (o *Outbound) getPeerForRequest(ctx context.Context, req *yarpc.Request) (*tchannelPeer, func(error), error) {
+	var peer yarpc.Peer
+	var onFinish func(error)
+	var err error
+	if o.Chooser != nil {
+		peer, onFinish, err = o.Chooser.Choose(ctx, req)
+		if err != nil {
+			return nil, nil, err
+		}
+	} else if o.Dialer != nil && o.Addr != "" {
+		peer, err = o.Dialer.RetainPeer(yarpc.Address(o.Addr), yarpc.NopSubscriber)
+		if err != nil {
+			return nil, nil, err
+		}
+		err = o.Dialer.ReleasePeer(yarpc.Address(o.Addr), yarpc.NopSubscriber)
+		if err != nil {
+			return nil, nil, err
+		}
+		onFinish = nopFinish
```

Comment:
Regardless of whether the user-facing API for `Chooser || (Dialer && Addr)` is
changed, I think we can handle it internally better by converting `Dialer &&
Addr` into a Chooser. It will simplify this code and give us the option of
doing that check once at Start time rather than per-request.

Reply:
Outbounds no longer have a Start method, and I suspect that we’ll be revisiting this pattern in all transports to send requests to call-site specified peers.